File: Bots/DotNet/Consumers/CodeFirst/WaterfallHostBot/Dialogs/MainDialog.cs
﻿// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.

using System;
using System.Collections.Generic;
using System.Linq;
using System.Threading;
using System.Threading.Tasks;
using Microsoft.Bot.Builder;
using Microsoft.Bot.Builder.Dialogs;
using Microsoft.Bot.Builder.Dialogs.Choices;
using Microsoft.Bot.Builder.Integration.AspNet.Core.Skills;
using Microsoft.Bot.Builder.Skills;
using Microsoft.Bot.Connector.Authentication;
using Microsoft.Bot.Schema;
using Microsoft.BotFrameworkFunctionalTests.WaterfallHostBot.Dialogs.Sso;
using Microsoft.Extensions.Configuration;
using Newtonsoft.Json;

namespace Microsoft.BotFrameworkFunctionalTests.WaterfallHostBot.Dialogs
{
    /// <summary>
    /// The main dialog for this bot. It uses a <see cref="SkillDialog"/> to call skills.
    /// </summary>
    public class MainDialog : ComponentDialog
    {
        // State property key that stores the active skill (used in AdapterWithErrorHandler to terminate the skills on error).
        public static readonly string ActiveSkillPropertyName = $"{typeof(MainDialog).FullName}.ActiveSkillProperty";

        // Constants used for selecting actions on the skill.
        private const string JustForwardTheActivity = "JustForwardTurnContext.Activity";
        
        private readonly IStatePropertyAccessor<BotFrameworkSkill> _activeSkillProperty;
        private readonly string _deliveryMode = $"{typeof(MainDialog).FullName}.DeliveryMode";
        private readonly string _selectedSkillKey = $"{typeof(MainDialog).FullName}.SelectedSkillKey";
        private readonly SkillsConfiguration _skillsConfig;

        // Dependency injection uses this constructor to instantiate MainDialog.
        public MainDialog(ConversationState conversationState, SkillConversationIdFactoryBase conversationIdFactory, SkillHttpClient skillClient, SkillsConfiguration skillsConfig, IConfiguration configuration)
            : base(nameof(MainDialog))
        {
            var botId = configuration.GetSection(MicrosoftAppCredentials.MicrosoftAppIdKey)?.Value;

            _skillsConfig = skillsConfig ?? throw new ArgumentNullException(nameof(skillsConfig));

            if (skillClient == null)
            {
                throw new ArgumentNullException(nameof(skillClient));
            }

            if (conversationState == null)
            {
                throw new ArgumentNullException(nameof(conversationState));
            }

            // Create state property to track the active skill.
            _activeSkillProperty = conversationState.CreateProperty<BotFrameworkSkill>(ActiveSkillPropertyName);

            // Register the tangent dialog for testing tangents and resume
            AddDialog(new TangentDialog());

            // Create and add SkillDialog instances for the configured skills.
            AddSkillDialogs(conversationState, conversationIdFactory, skillClient, skillsConfig, botId);

            // Add ChoicePrompt to render available delivery modes.
            AddDialog(new ChoicePrompt("DeliveryModePrompt"));

            // Add ChoicePrompt to render available types of skill.
            AddDialog(new ChoicePrompt("SkillTypePrompt"));

            // Add ChoicePrompt to render available skills.
            AddDialog(new ChoicePrompt("SkillPrompt"));

            // Add ChoicePrompt to render skill actions.
            AddDialog(new ChoicePrompt("SkillActionPrompt", SkillActionPromptValidator));

            // Add dialog to prepare SSO on the host and test the SSO skill
            // The waterfall skillDialog created in AddSkillDialogs contains the SSO skill action.
            var waterfallDialog = Dialogs
                .GetDialogs()
                .Where(e => e.Id.StartsWith("WaterfallSkill"))
                .First();
            AddDialog(new SsoDialog(waterfallDialog, configuration));

            // Add main waterfall dialog for this bot.
            var waterfallSteps = new WaterfallStep[]
            {
                SelectDeliveryModeStepAsync,
                SelectSkillTypeStepAsync,
                SelectSkillStepAsync,
                SelectSkillActionStepAsync,
                CallSkillActionStepAsync,
                FinalStepAsync
            };
            AddDialog(new WaterfallDialog(nameof(WaterfallDialog), waterfallSteps));
            
            // The initial child Dialog to run.
            InitialDialogId = nameof(WaterfallDialog);
        }

        /// <summary>
        /// This override is used to test the "abort" command to interrupt skills from the parent and
        /// also to test the "tangent" command to start a tangent and resume a skill.
        /// </summary>
        /// <param name="innerDc">The inner <see cref="DialogContext"/> for the current turn of conversation.</param>
        /// <param name="cancellationToken">A cancellation token that can be used by other objects
        /// or threads to receive notice of cancellation.</param>
        /// <returns>A <see cref="Task"/> representing the asynchronous operation.</returns>
        protected override async Task<DialogTurnResult> OnContinueDialogAsync(DialogContext innerDc, CancellationToken cancellationToken = default)
        {
            // This is an example on how to cancel a SkillDialog that is currently in progress from the parent bot.
            var activeSkill = await _activeSkillProperty.GetAsync(innerDc.Context, () => null, cancellationToken);
            var activity = innerDc.Context.Activity;
            if (activeSkill != null && activity.Type == ActivityTypes.Message && !string.IsNullOrWhiteSpace(activity.Text) && activity.Text.Equals("abort", StringComparison.CurrentCultureIgnoreCase))
            {
                // Cancel all dialogs when the user says abort.
                // The SkillDialog automatically sends an EndOfConversation message to the skill to let the
                // skill know that it needs to end its current dialogs, too.
                await innerDc.CancelAllDialogsAsync(cancellationToken);
                return await innerDc.ReplaceDialogAsync(InitialDialogId, "Canceled! \n\n What delivery mode would you like to use?", cancellationToken);
            }

            // Sample to test a tangent when in the middle of a skill conversation.
            if (activeSkill != null && activity.Type == ActivityTypes.Message && !string.IsNullOrWhiteSpace(activity.Text) && activity.Text.Equals("tangent", StringComparison.CurrentCultureIgnoreCase))
            {
                // Start tangent.
                return await innerDc.BeginDialogAsync(nameof(TangentDialog), cancellationToken: cancellationToken);
            }

            return await base.OnContinueDialogAsync(innerDc, cancellationToken);
        }

        // Render a prompt to select the delivery mode to use.
        private async Task<DialogTurnResult> SelectDeliveryModeStepAsync(WaterfallStepContext stepContext, CancellationToken cancellationToken)
        {
            // Create the PromptOptions with the delivery modes supported.
            var messageText = stepContext.Options?.ToString() ?? "What delivery mode would you like to use?";
            const string rePromptMessageText = "That was not a valid choice, please select a valid delivery mode.";
            var choices = new List<Choice>
            {
                new Choice(DeliveryModes.Normal),
                new Choice(DeliveryModes.ExpectReplies)
            };
            var options = new PromptOptions
            {
                Prompt = MessageFactory.Text(messageText, messageText, InputHints.ExpectingInput),
                RetryPrompt = MessageFactory.Text(rePromptMessageText, rePromptMessageText, InputHints.ExpectingInput),
                Style = ListStyle.SuggestedAction,
                Choices = choices
            };

            // Prompt the user to select a delivery mode.
            return await stepContext.PromptAsync("DeliveryModePrompt", options, cancellationToken);
        }

        // Render a prompt to select the type of skill to use.
        private async Task<DialogTurnResult> SelectSkillTypeStepAsync(WaterfallStepContext stepContext, CancellationToken cancellationToken)
        {
            // Remember the delivery mode selected by the user.
            stepContext.Values[_deliveryMode] = ((FoundChoice)stepContext.Result).Value;

            // Create the PromptOptions with the types of supported skills.
            const string messageText = "What type of skill would you like to use?";
            const string rePromptMessageText = "That was not a valid choice, please select a valid skill type.";
            var choices = new List<Choice>
            {
                new Choice("EchoSkill"),
                new Choice("WaterfallSkill")
            };
            var options = new PromptOptions
            {
                Prompt = MessageFactory.Text(messageText, messageText, InputHints.ExpectingInput),
                RetryPrompt = MessageFactory.Text(rePromptMessageText, rePromptMessageText, InputHints.ExpectingInput),
                Style = ListStyle.SuggestedAction,
                Choices = choices
            };

            // Prompt the user to select a type of skill.
            return await stepContext.PromptAsync("SkillTypePrompt", options, cancellationToken);
        }

        // Render a prompt to select the skill to call.
        private async Task<DialogTurnResult> SelectSkillStepAsync(WaterfallStepContext stepContext, CancellationToken cancellationToken)
        {
            var skillType = ((FoundChoice)stepContext.Result).Value;

            // Create the PromptOptions from the skill configuration which contain the list of configured skills.
            const string messageText = "What skill would you like to call?";
            const string repromptMessageText = "That was not a valid choice, please select a valid skill.";

            var choices = _skillsConfig.Skills
                .Where(skill => skill.Key.StartsWith(skillType))
                .Select(skill => new Choice(skill.Value.Id))
                .ToList();

            var options = new PromptOptions
            {
                Prompt = MessageFactory.Text(messageText, messageText, InputHints.ExpectingInput),
                RetryPrompt = MessageFactory.Text(repromptMessageText, repromptMessageText, InputHints.ExpectingInput),
                Style = ListStyle.SuggestedAction,
                Choices = choices
            };

            // Prompt the user to select a skill.
            return await stepContext.PromptAsync("SkillPrompt", options, cancellationToken);
        }

        // Render a prompt to select the begin action for the skill.
        private async Task<DialogTurnResult> SelectSkillActionStepAsync(WaterfallStepContext stepContext, CancellationToken cancellationToken)
        {
            // Get the skill info based on the selected skill.
            var selectedSkillId = ((FoundChoice)stepContext.Result).Value;
            var deliveryMode = stepContext.Values[_deliveryMode].ToString();
            var v3Bots = new List<string> { "EchoSkillBotDotNetV3", "EchoSkillBotJSV3" };

            // Exclude v3 bots from ExpectReplies
            if (deliveryMode == DeliveryModes.ExpectReplies && v3Bots.Contains(selectedSkillId))
            {
                await stepContext.Context.SendActivityAsync(MessageFactory.Text("V3 Bots do not support 'expectReplies' delivery mode."), cancellationToken);

                // Restart setup dialog
                return await stepContext.ReplaceDialogAsync(InitialDialogId, null, cancellationToken);
            }

            var selectedSkill = _skillsConfig.Skills.FirstOrDefault(keyValuePair => keyValuePair.Value.Id == selectedSkillId).Value;

            // Remember the skill selected by the user.
            stepContext.Values[_selectedSkillKey] = selectedSkill;

            // Create the PromptOptions with the actions supported by the selected skill.
            var messageText = $"Select an action # to send to **{selectedSkill.Id}**.\n\nOr just type in a message and it will be forwarded to the skill as a message activity.";
            var options = new PromptOptions
            {
                Prompt = MessageFactory.Text(messageText, messageText, InputHints.ExpectingInput),
                Choices = selectedSkill.GetActions().Select(action => new Choice(action)).ToList()
            };

            // Prompt the user to select a skill action.
            return await stepContext.PromptAsync("SkillActionPrompt", options, cancellationToken);
        }

        // This validator defaults to Message if the user doesn't select an existing option.
        private Task<bool> SkillActionPromptValidator(PromptValidatorContext<FoundChoice> promptContext, CancellationToken cancellationToken)
        {
            if (!promptContext.Recognized.Succeeded)
            {
                // Assume the user wants to send a message if an item in the list is not selected.
                promptContext.Recognized.Value = new FoundChoice { Value = JustForwardTheActivity };
            }

            return Task.FromResult(true);
        }

        // Starts the SkillDialog based on the user's selections.
        private async Task<DialogTurnResult> CallSkillActionStepAsync(WaterfallStepContext stepContext, CancellationToken cancellationToken)
        {
            var selectedSkill = (BotFrameworkSkill)stepContext.Values[_selectedSkillKey];

            var skillActivity = CreateBeginActivity(stepContext.Context, selectedSkill.Id, ((FoundChoice)stepContext.Result).Value);

            // Create the BeginSkillDialogOptions and assign the activity to send.
            var skillDialogArgs = new BeginSkillDialogOptions { Activity = skillActivity };

            var deliveryMode = stepContext.Values[_deliveryMode].ToString();

            if (deliveryMode == DeliveryModes.ExpectReplies)
            {
                skillDialogArgs.Activity.DeliveryMode = DeliveryModes.ExpectReplies;
            }

            // Save active skill in state.
            await _activeSkillProperty.SetAsync(stepContext.Context, selectedSkill, cancellationToken);

            if (skillActivity.Name == "Sso")
            {
                // Special case, we start the SSO dialog to prepare the host to call the skill.
                return await stepContext.BeginDialogAsync(nameof(SsoDialog), cancellationToken: cancellationToken);
            }

            // Start the skillDialog instance with the arguments. 
            return await stepContext.BeginDialogAsync(selectedSkill.Id, skillDialogArgs, cancellationToken);
        }

        // The SkillDialog has ended, render the results (if any) and restart MainDialog.
        private async Task<DialogTurnResult> FinalStepAsync(WaterfallStepContext stepContext, CancellationToken cancellationToken)
        {
            var activeSkill = await _activeSkillProperty.GetAsync(stepContext.Context, () => null, cancellationToken);

            // Check if the skill returned any results and display them.
            if (stepContext.Result != null)
            {
                var message = $"Skill \"{activeSkill.Id}\" invocation complete.";
                message += $" Result: {JsonConvert.SerializeObject(stepContext.Result)}";
                await stepContext.Context.SendActivityAsync(MessageFactory.Text(message, message, inputHint: InputHints.IgnoringInput), cancellationToken: cancellationToken);
            }

            // Clear the delivery mode selected by the user.
            stepContext.Values[_deliveryMode] = null;

            // Clear the skill selected by the user.
            stepContext.Values[_selectedSkillKey] = null;

            // Clear active skill in state.
            await _activeSkillProperty.DeleteAsync(stepContext.Context, cancellationToken);

            // Restart the main dialog with a different message the second time around.
            return await stepContext.ReplaceDialogAsync(InitialDialogId, $"Done with \"{activeSkill.Id}\". \n\n What delivery mode would you like to use?", cancellationToken);
        }

        // Helper method that creates and adds SkillDialog instances for the configured skills.
        private void AddSkillDialogs(ConversationState conversationState, SkillConversationIdFactoryBase conversationIdFactory, SkillHttpClient skillClient, SkillsConfiguration skillsConfig, string botId)
        {
            foreach (var skillInfo in _skillsConfig.Skills.Values)
            {
                // Create the dialog options.
                var skillDialogOptions = new SkillDialogOptions
                {
                    BotId = botId,
                    ConversationIdFactory = conversationIdFactory,
                    SkillClient = skillClient,
                    SkillHostEndpoint = skillsConfig.SkillHostEndpoint,
                    ConversationState = conversationState,
                    Skill = skillInfo
                };

                // Add a SkillDialog for the selected skill.
                AddDialog(new SkillDialog(skillDialogOptions, skillInfo.Id));
            }
        }

        // Helper method to create the activity to be sent to the DialogSkillBot using selected type and values.
        private Activity CreateBeginActivity(ITurnContext turnContext, string skillId, string selectedOption)
        {
            if (selectedOption.Equals(JustForwardTheActivity, StringComparison.CurrentCultureIgnoreCase))
            {
                // Note message activities also support input parameters but we are not using them in this example.
                // Return a deep clone of the activity so we don't risk altering the original one 
                return ObjectPath.Clone(turnContext.Activity);
            }

            // Get the begin activity from the skill instance.
            var activity = _skillsConfig.Skills[skillId].CreateBeginActivity(selectedOption);

            // We are manually creating the activity to send to the skill; ensure we add the ChannelData and Properties 
            // from the original activity so the skill gets them.
            // Note: this is not necessary if we are just forwarding the current activity from context. 
            activity.ChannelData = turnContext.Activity.ChannelData;
            activity.Properties = turnContext.Activity.Properties;

            return activity;
        }
    }
}

File: Bots/Python/Consumers/CodeFirst/WaterfallHostBot/dialogs/main_dialog.py
# Copyright (c) Microsoft Corporation. All rights reserved.
# Licensed under the MIT License.

import copy
import json

from botbuilder.dialogs import (
    ComponentDialog,
    DialogContext,
    WaterfallDialog,
    WaterfallStepContext,
    DialogTurnResult,
    ListStyle,
)
from botbuilder.dialogs.choices import Choice, FoundChoice
from botbuilder.dialogs.prompts import (
    PromptOptions,
    ChoicePrompt,
    PromptValidatorContext,
)
from botbuilder.dialogs.skills import (
    SkillDialogOptions,
    SkillDialog,
    BeginSkillDialogOptions,
)
from botbuilder.core import ConversationState, MessageFactory, TurnContext
from botbuilder.core.skills import ConversationIdFactoryBase
from botbuilder.schema import Activity, ActivityTypes, InputHints, DeliveryModes
from botbuilder.integration.aiohttp.skills import SkillHttpClient

from skills_configuration import SkillsConfiguration, DefaultConfig

from dialogs.tangent_dialog import TangentDialog
from dialogs.sso.sso_dialog import SsoDialog


ACTIVE_SKILL_PROPERTY_NAME = "MainDialog.ActiveSkillProperty"
DELIVERY_MODE_NAME = "MainDialog.DeliveryMode"
SELECTED_SKILL_KEY_NAME = "MainDialog.SelectedSkillKey"
JUST_FORWARD_THE_ACTIVITY = "JustForwardTurnContext.Activity"

DELIVERY_MODE_PROMPT = "DeliveryModePrompt"
SKILL_TYPE_PROMPT = "SkillTypePrompt"
SKILL_PROMPT = "SkillPrompt"
SKILL_ACTION_PROMPT = "SkillActionPrompt"


class MainDialog(ComponentDialog):
    def __init__(
        self,
        conversation_state: ConversationState,
        conversation_id_factory: ConversationIdFactoryBase,
        skill_client: SkillHttpClient,
        skills_config: SkillsConfiguration,
        configuration: DefaultConfig,
    ):
        super().__init__(MainDialog.__name__)

        bot_id = configuration.APP_ID

        self._skills_config = skills_config
        if not self._skills_config:
            raise TypeError(
                "[MainDialog]: Missing parameter. skills_config is required"
            )

        if not skill_client:
            raise TypeError("[MainDialog]: Missing parameter. skill_client is required")

        if not conversation_state:
            raise TypeError(
                "[MainDialog]: Missing parameter. conversation_state is required"
            )

        if not conversation_id_factory:
            raise TypeError(
                "[MainDialog]: Missing parameter. conversation_id_factory is required"
            )

        # Use helper method to add SkillDialog instances for the configured skills.
        self._add_skill_dialogs(
            conversation_state,
            conversation_id_factory,
            skill_client,
            skills_config,
            bot_id,
        )

        # Create state property to track the active skill.
        self.active_skill_property = conversation_state.create_property(
            ACTIVE_SKILL_PROPERTY_NAME
        )

        # Register the tangent dialog for testing tangents and resume
        self.add_dialog(TangentDialog())

        # Add ChoicePrompt to render available delivery modes.
        self.add_dialog(ChoicePrompt(DELIVERY_MODE_PROMPT))

        # Add ChoicePrompt to render available types of skill.
        self.add_dialog(ChoicePrompt(SKILL_TYPE_PROMPT))

        # Add ChoicePrompt to render available skills.
        self.add_dialog(ChoicePrompt(SKILL_PROMPT))

        # Add ChoicePrompt to render skill actions.
        self.add_dialog(
            ChoicePrompt(SKILL_ACTION_PROMPT, self._skill_action_prompt_validator)
        )

        # Add dialog to prepare SSO on the host and test the SSO skill
        # The waterfall skillDialog created in AddSkillDialogs contains the SSO skill action.
        waterfall_skills = [
            skill_dialog
            for skill_dialog in self._dialogs._dialogs.values()
            if skill_dialog.id.startswith("WATERFALLSKILL")
        ]

        for waterfall_skill in waterfall_skills:
            self.add_dialog(SsoDialog(configuration, waterfall_skill))

        self.add_dialog(
            WaterfallDialog(
                WaterfallDialog.__name__,
                [
                    self._select_delivery_mode_step,
                    self._select_skill_type_step,
                    self._select_skill_step,
                    self._select_skill_action_step,
                    self._call_skill_action_step,
                    self._final_step,
                ],
            )
        )

        self.initial_dialog_id = WaterfallDialog.__name__

    async def on_continue_dialog(self, inner_dc: DialogContext):
        """
        This override is used to test the "abort" command to interrupt skills from the parent and
        also to test the "tangent" command to start a tangent and resume a skill.
        """

        # This is an example on how to cancel a SkillDialog that is currently in progress from the parent bot.
        active_skill = await self.active_skill_property.get(inner_dc.context)
        activity = inner_dc.context.activity

        if (
            active_skill
            and activity.type == ActivityTypes.message
            and "abort" in activity.text.lower()
        ):
            # Cancel all dialogs when the user says abort.
            # The SkillDialog automatically sends an EndOfConversation message to the skill to let the
            # skill know that it needs to end its current dialogs, too.
            await inner_dc.cancel_all_dialogs()
            return await inner_dc.replace_dialog(
                self.initial_dialog_id,
                "Canceled! \n\n What delivery mode would you like to use?",
            )

        # Sample to test a tangent when in the middle of a skill conversation.
        if (
            active_skill
            and activity.type == ActivityTypes.message
            and activity.text
            and activity.text.lower() == "tangent"
        ):
            # Start tangent.
            return await inner_dc.begin_dialog(TangentDialog.__name__)

        return await super().on_continue_dialog(inner_dc)

    async def _select_delivery_mode_step(
        self, step_context: WaterfallStepContext
    ) -> DialogTurnResult:
        """
        Render a prompt to select the delivery mode to use.
        """

        # Create the PromptOptions with the delivery modes supported.
        message_text = (
            str(step_context.options)
            if step_context.options
            else "What delivery mode would you like to use?"
        )
        reprompt_message_text = (
            "That was not a valid choice, please select a valid delivery mode."
        )

        options = PromptOptions(
            prompt=MessageFactory.text(
                message_text, message_text, InputHints.expecting_input
            ),
            retry_prompt=MessageFactory.text(
                reprompt_message_text, reprompt_message_text, InputHints.expecting_input
            ),
            style=ListStyle.suggested_action,
            choices=[
                Choice(DeliveryModes.normal.value),
                Choice(DeliveryModes.expect_replies.value),
            ],
        )

        # Prompt the user to select a delivery mode.
        return await step_context.prompt(DELIVERY_MODE_PROMPT, options)

    async def _select_skill_type_step(self, step_context: WaterfallStepContext):
        """
        Render a prompt to select the type of skill to use.
        """

        # Remember the delivery mode selected by the user.
        step_context.values[DELIVERY_MODE_NAME] = step_context.result.value

        # Create the PromptOptions with the types of supported skills.
        message_text = "What type of skill would you like to use?"
        reprompt_message_text = (
            "That was not a valid choice, please select a valid skill type."
        )

        options = PromptOptions(
            prompt=MessageFactory.text(
                message_text, message_text, InputHints.expecting_input
            ),
            retry_prompt=MessageFactory.text(
                reprompt_message_text, reprompt_message_text, InputHints.expecting_input
            ),
            style=ListStyle.suggested_action,
            choices=[Choice("EchoSkill"), Choice("WaterfallSkill")],
        )

        # Prompt the user to select a type of skill.
        return await step_context.prompt(SKILL_TYPE_PROMPT, options)

    async def _select_skill_step(
        self, step_context: WaterfallStepContext
    ) -> DialogTurnResult:
        """
        Render a prompt to select the skill to call.
        """

        skill_type = step_context.result.value

        # Create the PromptOptions from the skill configuration which contain the list of configured skills.
        message_text = "What skill would you like to call?"
        reprompt_message_text = (
            "That was not a valid choice, please select a valid skill."
        )

        options = PromptOptions(
            prompt=MessageFactory.text(
                message_text, message_text, InputHints.expecting_input
            ),
            retry_prompt=MessageFactory.text(
                reprompt_message_text, reprompt_message_text, InputHints.expecting_input
            ),
            style=ListStyle.suggested_action,
            choices=[
                Choice(val.id)
                for key, val in self._skills_config.SKILLS.items()
                if key.lower().startswith(skill_type.lower())
            ],
        )

        # Prompt the user to select a skill.
        return await step_context.prompt(SKILL_PROMPT, options)

    async def _select_skill_action_step(self, step_context: WaterfallStepContext):
        """
        Render a prompt to select the begin action for the skill.
        """

        # Get the skill info based on the selected skill.
        selected_skill_id = step_context.result.value
        delivery_mode = str(step_context.values[DELIVERY_MODE_NAME])
        v3_bots = ["EchoSkillBotDotNetV3", "EchoSkillBotJSV3"]

        # Exclude v3 bots from ExpectReplies
        if (
            delivery_mode == DeliveryModes.expect_replies
            and selected_skill_id in v3_bots
        ):
            await step_context.context.send_activity(
                MessageFactory.text(
                    "V3 Bots do not support 'expectReplies' delivery mode."
                )
            )

            # Restart setup dialog
            return await step_context.replace_dialog(self.initial_dialog_id)

        selected_skill = next(
            val
            for key, val in self._skills_config.SKILLS.items()
            if val.id == selected_skill_id
        )

        # Remember the skill selected by the user.
        step_context.values[SELECTED_SKILL_KEY_NAME] = selected_skill

        # Create the PromptOptions with the actions supported by the selected skill.
        message_text = (
            f"Select an action # to send to **{selected_skill.id}**.\n\nOr just type in "
            f"a message and it will be forwarded to the skill as a message activity."
        )

        options = PromptOptions(
            prompt=MessageFactory.text(
                message_text, message_text, InputHints.expecting_input
            ),
            choices=[Choice(action) for action in selected_skill.get_actions()],
        )

        # Prompt the user to select a skill action.
        return await step_context.prompt(SKILL_ACTION_PROMPT, options)

    async def _skill_action_prompt_validator(
        self, prompt_context: PromptValidatorContext
    ):
        """
        This validator defaults to Message if the user doesn't select an existing option.
        """

        if not prompt_context.recognized.succeeded:
            # Assume the user wants to send a message if an item in the list is not selected.
            prompt_context.recognized.value = FoundChoice(
                value=JUST_FORWARD_THE_ACTIVITY,
                index=0,
                score=0,
            )

        return True

    async def _call_skill_action_step(self, step_context: WaterfallStepContext):
        """
        Starts the SkillDialog based on the user's selections.
        """

        selected_skill = step_context.values[SELECTED_SKILL_KEY_NAME]
        skill_activity = self._create_begin_activity(
            step_context.context, selected_skill.id, step_context.result.value
        )

        # Create the BeginSkillDialogOptions and assign the activity to send.
        skill_dialog_args = BeginSkillDialogOptions(activity=skill_activity)
        delivery_mode = str(step_context.values[DELIVERY_MODE_NAME])

        if delivery_mode == DeliveryModes.expect_replies:
            skill_dialog_args.activity.delivery_mode = DeliveryModes.expect_replies

        # Save active skill in state.
        await self.active_skill_property.set(step_context.context, selected_skill)

        if skill_activity.name == "Sso":
            # Special case, we start the SSO dialog to prepare the host to call the skill.
            return await step_context.begin_dialog(
                SsoDialog.__name__ + selected_skill.id
            )

        # Start the skillDialog instance with the arguments.
        return await step_context.begin_dialog(selected_skill.id, skill_dialog_args)

    async def _final_step(self, step_context: WaterfallStepContext):
        """
        The SkillDialog has ended, render the results (if any) and restart MainDialog.
        """

        active_skill = await self.active_skill_property.get(step_context.context)

        # Check if the skill returned any results and display them.
        if step_context.result:
            message = f'Skill "{active_skill.id}" invocation complete.'
            message += f" Result: {json.dumps(step_context.result)}"
            await step_context.context.send_activity(
                MessageFactory.text(message, message, InputHints.ignoring_input)
            )

        # Clear the delivery mode selected by the user.
        step_context.values[DELIVERY_MODE_NAME] = None

        # Clear the skill selected by the user.
        step_context.values[SELECTED_SKILL_KEY_NAME] = None

        # Clear active skill in state.
        await self.active_skill_property.delete(step_context.context)

        # Restart the main dialog with a different message the second time around.
        return await step_context.replace_dialog(
            self.initial_dialog_id,
            f'Done with "{active_skill.id}". \n\n What delivery mode would you '
            f"like to use?",
        )

    def _add_skill_dialogs(
        self,
        conversation_state: ConversationState,
        conversation_id_factory: ConversationIdFactoryBase,
        skill_client: SkillHttpClient,
        skills_config: SkillsConfiguration,
        bot_id: str,
    ):
        """
        Helper method that creates and adds SkillDialog instances for the configured skills.
        """

        for skill_info in self._skills_config.SKILLS.values():
            # Create the dialog options.
            skill_dialog_options = SkillDialogOptions(
                bot_id=bot_id,
                conversation_id_factory=conversation_id_factory,
                skill_client=skill_client,
                skill_host_endpoint=skills_config.SKILL_HOST_ENDPOINT,
                conversation_state=conversation_state,
                skill=skill_info,
            )

            # Add a SkillDialog for the selected skill.
            self.add_dialog(SkillDialog(skill_dialog_options, skill_info.id))

    def _create_begin_activity(
        self, context: TurnContext, skill_id: str, selected_option: str
    ):
        if selected_option.lower() == JUST_FORWARD_THE_ACTIVITY.lower():
            # Note message activities also support input parameters but we are not using them in this example.
            # Return a deep clone of the activity so we don't risk altering the original one
            return copy.deepcopy(context.activity)

        # Get the begin activity from the skill instance.
        activity: Activity = self._skills_config.SKILLS[skill_id].create_begin_activity(
            selected_option
        )

        # We are manually creating the activity to send to the skill; ensure we add the ChannelData and Properties
        # from the original activity so the skill gets them.
        # Note: this is not necessary if we are just forwarding the current activity from context.
        activity.channel_data = context.activity.channel_data
        activity.additional_properties = context.activity.additional_properties

        return activity
